syntax = "proto3";

package raft;

service Raft {
    rpc RequestVote(RequestVoteArgs) returns (RequestVoteReply);
    rpc AppendEntries(AppendEntriesArgs) returns (AppendEntriesReply);
    rpc SendResponse(ResponseMessage) returns (ResponseAck);
}

message RequestVoteArgs {
    int32 term = 1;
    int32 candidateId = 2;
    int32 lastLogIndex = 3;
    int32 lastLogTerm = 4;
}

message RequestVoteReply {
    int32 term = 1;
    bool voteGranted = 2;
}

message AppendEntriesArgs {
    int32 term = 1;
    int32 leaderId = 2;
    repeated string entries = 3;
}

message AppendEntriesReply {
    int32 term = 1;
    bool success = 2;
}

message ResponseMessage {
    int32 senderId = 1;
    string message = 2;
}

message ResponseAck {
    bool success = 1;
}
